example/inline/composition.yaml

```diff
@@ -26,11 +26,7 @@ spec:
                 {{ setResourceNameAnnotation (print "test-user-" $i) }}
               labels:
                 testing.upbound.io/example-name: test-user-{{ $i }}
-              {{ if eq $.observed.resources nil }}
-                dummy: {{ randomChoice "foo" "bar" "baz" }}
-              {{ else }}
-                dummy: {{ ( index $.observed.resources ( print "test-user-" $i ) ).resource.metadata.labels.dummy }}
-              {{ end }}
+                dummy: {{ default ( randomChoice "foo" "bar" "baz" ) ( get $.observed.resources ( print "test-user-" $i ) ) }}
```

Comment:
i tested this logic and I think it may not be fully correct still.  If the resource already exists, are we trying to assign the full content of the entire `test-user-i` resource to a label? The event below seems to imply we're trying to write an entire resource there as opposed to just a string value for the label. What do you think @ezgidemirel? 🤔 

Full event details:
```
Warning  ComposeResources         42s (x6 over 3m42s)  defined/compositeresourcedefinition.apiextensions.crossplane.io  cannot compose resources: cannot apply composed resource "test-user-0": User.iam.aws.upbound.io "example-pdnsb-sd8jw" is invalid: [metadata.labels: Invalid value: "map[resource:map[apiVersion:iam.aws.upbound.io/v1beta1 kind:User metadata:map[annotations:map[crossplane.io/composition-resource-name:test-user-0 crossplane.io/external-create-pending:2023-12-11T17:55:19Z crossplane.io/external-create-succeeded:2023-12-11T17:55:19Z crossplane.io/external-name:example-pdnsb-sd8jw] creationTimestamp:2023-12-11T17:55:18Z finalizers:[finalizer.managedresource.crossplane.io] generateName:example-pdnsb- generation:3 labels:map[crossplane.io/claim-name:example crossplane.io/claim-namespace:default crossplane.io/composite:example-pdnsb dummy:bar testing.upbound.io/example-name:test-user-0] managedFields:[map[apiVersion:iam.aws.upbound.io/v1beta1 fieldsType:FieldsV1 fieldsV1:map[f:metadata:map[f:annotations:map[f:crossplane.io/composition-resource-name:map[]] f:generateName:map[] f:labels:map[f:crossplane.io/claim-name:map[] f:crossplane.io/claim-namespace:map[] f:crossplane.io/composite:map[] f:dummy:map[] f:testing.upbound.io/example-name:map[]] f:ownerReferences:map[k:{\"uid\":\"19c2e541-d3ce-49b6-a327-314105add4f0\"}:map[]]] f:spec:map[f:forProvider:map[]]] manager:apiextensions.crossplane.io/composed operation:Apply time:2023-12-11T17:55:18Z] map[apiVersion:iam.aws.upbound.io/v1beta1 fieldsType:FieldsV1 fieldsV1:map[f:metadata:map[f:annotations:map[f:crossplane.io/external-create-pending:map[] f:crossplane.io/external-create-succeeded:map[] f:crossplane.io/external-name:map[]] f:finalizers:map[.:map[] v:\"finalizer.managedresource.crossplane.io\":map[]]] f:spec:map[f:forProvider:map[f:path:map[] f:tags:map[.:map[] f:crossplane-kind:map[] f:crossplane-name:map[] f:crossplane-providerconfig:map[]]] f:initProvider:map[]]] manager:provider operation:Update time:2023-12-11T17:55:19Z] map[apiVersion:iam.aws.upbound.io/v1beta1 fieldsType:FieldsV1 fieldsV1:map[f:status:map[.:map[] f:atProvider:map[.:map[] f:arn:map[] f:forceDestroy:map[] f:id:map[] f:path:map[] f:tags:map[.:map[] f:crossplane-kind:map[] f:crossplane-name:map[] f:crossplane-providerconfig:map[]] f:tagsAll:map[.:map[] f:crossplane-kind:map[] f:crossplane-name:map[] f:crossplane-providerconfig:map[]] f:uniqueId:map[]] f:conditions:map[.:map[] k:{\"type\":\"LastAsyncOperation\"}:map[.:map[] f:lastTransitionTime:map[] f:reason:map[] f:status:map[] f:type:map[]] k:{\"type\":\"Ready\"}:map[.:map[] f:lastTransitionTime:map[] f:reason:map[] f:status:map[] f:type:map[]] k:{\"type\":\"Synced\"}:map[.:map[] f:lastTransitionTime:map[] f:reason:map[] f:status:map[] f:type:map[]]]]] manager:provider operation:Update subresource:status time:2023-12-11T17:55:19Z]] name:example-pdnsb-sd8jw ownerReferences:[map[apiVersion:example.crossplane.io/v1beta1 blockOwnerDeletion:true controller:true kind:UserAccess name:example-pdnsb uid:19c2e541-d3ce-49b6-a327-314105add4f0]] resourceVersion:432689 uid:1c274d5f-844b-473c-832a-93b19cd28829] spec:map[deletionPolicy:Delete forProvider:map[path:/ tags:map[crossplane-kind:user.iam.aws.upbound.io crossplane-name:example-pdnsb-sd8jw crossplane-providerconfig:default]] initProvider:map[] managementPolicies:[*] providerConfigRef:map[name:default]] status:map[atProvider:map[arn:arn:aws:iam::609897127049:user/example-pdnsb-sd8jw forceDestroy:false id:example-pdnsb-sd8jw path:/ tags:map[crossplane-kind:user.iam.aws.upbound.io crossplane-name:example-pdnsb-sd8jw crossplane-providerconfig:default] tagsAll:map[crossplane-kind:user.iam.aws.upbound.io crossplane-name:example-pdnsb-sd8jw crossplane-providerconfig:default] uniqueId:AIDAY4AFTTSERFM5NU32Y] conditions:[map[lastTransitionTime:2023-12-11T17:55:19Z reason:Available status:True type:Ready] map[lastTransitionTime:2023-12-11T17:55:19Z reason:ReconcileSuccess status:True type:Synced] map[lastTransitionTime:2023-12-11T17:55:19Z reason:Success status:True type:LastAsyncOperation]]]]]": must be no more than 63 characters, metadata.labels: Invalid value: "map[resource:map[apiVersion:iam.aws.upbound.io/v1beta1 kind:User metadata:map[annotations:map[crossplane.io/composition-resource-name:test-user-0 crossplane.io/external-create-pending:2023-12-11T17:55:19Z crossplane.io/external-create-succeeded:2023-12-11T17:55:19Z crossplane.io/external-name:example-pdnsb-sd8jw] creationTimestamp:2023-12-11T17:55:18Z finalizers:[finalizer.managedresource.crossplane.io] generateName:example-pdnsb- generation:3 labels:map[crossplane.io/claim-name:example crossplane.io/claim-namespace:default crossplane.io/composite:example-pdnsb dummy:bar testing.upbound.io/example-name:test-user-0] managedFields:[map[apiVersion:iam.aws.upbound.io/v1beta1 fieldsType:FieldsV1 fieldsV1:map[f:metadata:map[f:annotations:map[f:crossplane.io/composition-resource-name:map[]] f:generateName:map[] f:labels:map[f:crossplane.io/claim-name:map[] f:crossplane.io/claim-namespace:map[] f:crossplane.io/composite:map[] f:dummy:map[] f:testing.upbound.io/example-name:map[]] f:ownerReferences:map[k:{\"uid\":\"19c2e541-d3ce-49b6-a327-314105add4f0\"}:map[]]] f:spec:map[f:forProvider:map[]]] manager:apiextensions.crossplane.io/composed operation:Apply time:2023-12-11T17:55:18Z] map[apiVersion:iam.aws.upbound.io/v1beta1 fieldsType:FieldsV1 fieldsV1:map[f:metadata:map[f:annotations:map[f:crossplane.io/external-create-pending:map[] f:crossplane.io/external-create-succeeded:map[] f:crossplane.io/external-name:map[]] f:finalizers:map[.:map[] v:\"finalizer.managedresource.crossplane.io\":map[]]] f:spec:map[f:forProvider:map[f:path:map[] f:tags:map[.:map[] f:crossplane-kind:map[] f:crossplane-name:map[] f:crossplane-providerconfig:map[]]] f:initProvider:map[]]] manager:provider operation:Update time:2023-12-11T17:55:19Z] map[apiVersion:iam.aws.upbound.io/v1beta1 fieldsType:FieldsV1 fieldsV1:map[f:status:map[.:map[] f:atProvider:map[.:map[] f:arn:map[] f:forceDestroy:map[] f:id:map[] f:path:map[] f:tags:map[.:map[] f:crossplane-kind:map[] f:crossplane-name:map[] f:crossplane-providerconfig:map[]] f:tagsAll:map[.:map[] f:crossplane-kind:map[] f:crossplane-name:map[] f:crossplane-providerconfig:map[]] f:uniqueId:map[]] f:conditions:map[.:map[] k:{\"type\":\"LastAsyncOperation\"}:map[.:map[] f:lastTransitionTime:map[] f:reason:map[] f:status:map[] f:type:map[]] k:{\"type\":\"Ready\"}:map[.:map[] f:lastTransitionTime:map[] f:reason:map[] f:status:map[] f:type:map[]] k:{\"type\":\"Synced\"}:map[.:map[] f:lastTransitionTime:map[] f:reason:map[] f:status:map[] f:type:map[]]]]] manager:provider operation:Update subresource:status time:2023-12-11T17:55:19Z]] name:example-pdnsb-sd8jw ownerReferences:[map[apiVersion:example.crossplane.io/v1beta1 blockOwnerDeletion:true controller:true kind:UserAccess name:example-pdnsb uid:19c2e541-d3ce-49b6-a327-314105add4f0]] resourceVersion:432689 uid:1c274d5f-844b-473c-832a-93b19cd28829] spec:map[deletionPolicy:Delete forProvider:map[path:/ tags:map[crossplane-kind:user.iam.aws.upbound.io crossplane-name:example-pdnsb-sd8jw crossplane-providerconfig:default]] initProvider:map[] managementPolicies:[*] providerConfigRef:map[name:default]] status:map[atProvider:map[arn:arn:aws:iam::609897127049:user/example-pdnsb-sd8jw forceDestroy:false id:example-pdnsb-sd8jw path:/ tags:map[crossplane-kind:user.iam.aws.upbound.io crossplane-name:example-pdnsb-sd8jw crossplane-providerconfig:default] tagsAll:map[crossplane-kind:user.iam.aws.upbound.io crossplane-name:example-pdnsb-sd8jw crossplane-providerconfig:default] uniqueId:AIDAY4AFTTSERFM5NU32Y] conditions:[map[lastTransitionTime:2023-12-11T17:55:19Z reason:Available status:True type:Ready] map[lastTransitionTime:2023-12-11T17:55:19Z reason:ReconcileSuccess status:True type:Synced] map[lastTransitionTime:2023-12-11T17:55:19Z reason:Success status:True type:LastAsyncOperation]]]]]": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue',  or 'my_value',  or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')]
```

Reply:
Thanks for testing and catching the error @jbw976! I was just checking the label values, and making sure that they're not changing in each reconciliation. I wasn't paying attention to the events and missed the apply errors. This is a really useful lesson for me while testing/writing e2e scenarios :)  

I needed to switch from the `get` function to `dig` function. Because the `get` function returns a string. More details can be found in the sprig [doc](https://masterminds.github.io/sprig/dicts.html).